File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/utilities/ghostboxutility/views/WaveformView.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.utilities.ghostboxutility.views;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.ColorUtils;

/**
 * WaveformView class
 *
 * @author TritiumGamingStudios
 */
public class WaveformView extends View {
    @Nullable
    private byte[] mBytes;
    private float[] mPoints;

    private final float sampleRate = 128.0f;
    private final float peakAmplitude = .8f;

    private double xSpace = 0;
    private final double ySpace = .1;

    private final Rect waveformRect = new Rect();
    private final Rect frameRect = new Rect();

    private final Paint waveformPaint = new Paint();
    private final Paint framePaint = new Paint();
    private final Paint gridPaint = new Paint();
    private final Paint bgPaint = new Paint();

    /**
     * @param context
     */
    public WaveformView(Context context) {
        super(context);
        init();
    }

    /**
     * @param context
     * @param attrs
     */
    public WaveformView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    /**
     * @param context
     * @param attrs
     * @param defStyleAttr
     */
    public WaveformView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    /**
     * @param context
     * @param attrs
     * @param defStyleAttr
     * @param defStyleRes
     */
    public WaveformView(
            Context context,
            @Nullable AttributeSet attrs,
            int defStyleAttr,
            int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    /**
     *
     */
    public void init() {
        mBytes = null;

        @ColorInt int color_frame = Color.WHITE,
                color_grid = Color.argb(50, 150, 0, 0),
                color_bg = Color.argb(50, 200, 0, 0),
                color_waveform = Color.RED;

        color_waveform = ColorUtils.getColorFromAttribute(getContext(), R.attr.textColorPrimary);
        color_frame = ColorUtils.getColorFromAttribute(getContext(), R.attr.textColorSecondary);

        waveformPaint.setStrokeWidth(5f);
        waveformPaint.setStyle(Paint.Style.STROKE);
        waveformPaint.setAntiAlias(true);
        waveformPaint.setColor(color_waveform);

        framePaint.setStrokeWidth(10f);
        framePaint.setStyle(Paint.Style.STROKE);
        framePaint.setAntiAlias(true);
        framePaint.setColor(color_frame);

        gridPaint.setStrokeWidth(1f);
        gridPaint.setStyle(Paint.Style.STROKE);
        gridPaint.setAntiAlias(true);
        gridPaint.setColor(color_grid);

        bgPaint.setStrokeWidth(1f);
        bgPaint.setStyle(Paint.Style.FILL);
        bgPaint.setAntiAlias(true);
        bgPaint.setColor(color_bg);

    }

    /**
     * @param bytes
     */
    public void updateVisualizer(byte[] bytes) {
        this.mBytes = bytes;

        invalidate();
    }

    /**
     * @param canvas
     */
    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);

        xSpace = ySpace * ((double) getHeight() / (double) getWidth());

        frameRect.set(0, 0, getWidth(), getHeight());
        waveformRect.set(0, (int) (frameRect.height() * .2), frameRect.width(),
                (int) (frameRect.height() * .7));

        canvas.drawRect(frameRect, bgPaint);

        for (int y = 0; y < frameRect.height(); y += (int) (frameRect.height() * ySpace)) {
            canvas.drawLine(0, y, frameRect.width(), y, gridPaint);
        }
        for (int x = 0; x < frameRect.width(); x += (int) (frameRect.width() * xSpace)) {
            canvas.drawLine(x, 0, x, frameRect.height(), gridPaint);
        }

        if (mBytes == null) {
            canvas.drawLine(
                    0, waveformRect.height(),
                    waveformRect.width(), waveformRect.height(),
                    waveformPaint);
            return;
        }

        if (mPoints == null || mPoints.length < mBytes.length * 4) {
            mPoints = new float[mBytes.length * 4];
        }

        for (int i = 0; i < mBytes.length - 1; i++) {
            mPoints[i * 4] = (float) waveformRect.width() * i / (float) (mBytes.length - 1);
            mPoints[i * 4 + 1] = (float) waveformRect.height() + ((byte) (mBytes[i] + sampleRate)) * (waveformRect.height() * peakAmplitude) / sampleRate;
            mPoints[i * 4 + 2] = (float) waveformRect.width() * (i + 1) / ((float) mBytes.length - 1);
            mPoints[i * 4 + 3] = (float) waveformRect.height() + ((byte) (mBytes[i + 1] + sampleRate)) * (waveformRect.height() * peakAmplitude) / sampleRate;
        }

        canvas.drawLines(mPoints, waveformPaint);
    }
}

File: app-legacy/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/activities/investigation/utilities/ghostboxutility/views/WaveformView.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.utilities.ghostboxutility.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.view.View
import androidx.annotation.ColorInt
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.util.ColorUtils.getColorFromAttribute

/**
 * WaveformView class
 *
 * @author TritiumGamingStudios
 */
class WaveformView : View {
    private var mBytes: ByteArray? = null
    private var mPoints: FloatArray? = null

    private val sampleRate = 128.0f
    private val peakAmplitude = .8f

    private var xSpace = 0.0
    private val ySpace = .1

    private val waveformRect = Rect()
    private val frameRect = Rect()

    private val waveformPaint = Paint()
    private val framePaint = Paint()
    private val gridPaint = Paint()
    private val bgPaint = Paint()

    constructor(context: Context?) : super(context) { init() }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) { init() }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr) { init() }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) :
            super(context, attrs, defStyleAttr, defStyleRes) { init() }

    fun init() {
        mBytes = null
        @ColorInt val gridColor = Color.argb(50, 150, 0, 0)
        @ColorInt val backgroundColor = Color.argb(50, 200, 0, 0)

        @ColorInt val waveformColor = getColorFromAttribute(context, R.attr.textColorPrimary)
        @ColorInt val frameColor = getColorFromAttribute(context, R.attr.textColorSecondary)

        waveformPaint.strokeWidth = 5f
        waveformPaint.style = Paint.Style.STROKE
        waveformPaint.isAntiAlias = true
        waveformPaint.color = waveformColor

        framePaint.strokeWidth = 10f
        framePaint.style = Paint.Style.STROKE
        framePaint.isAntiAlias = true
        framePaint.color = frameColor

        gridPaint.strokeWidth = 1f
        gridPaint.style = Paint.Style.STROKE
        gridPaint.isAntiAlias = true
        gridPaint.color = gridColor

        bgPaint.strokeWidth = 1f
        bgPaint.style = Paint.Style.FILL
        bgPaint.isAntiAlias = true
        bgPaint.color = backgroundColor
    }

    fun updateVisualizer(bytes: ByteArray?) {
        this.mBytes = bytes
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        xSpace = ySpace * (height.toDouble() / width.toDouble())

        frameRect[0, 0, width] = height
        waveformRect[0, (frameRect.height() * .2).toInt(), frameRect.width()] =
            (frameRect.height() * .7).toInt()

        canvas.drawRect(frameRect, bgPaint)

        var y = 0
        while (y < frameRect.height()) {
            canvas.drawLine(0f, y.toFloat(), frameRect.width().toFloat(), y.toFloat(), gridPaint)
            y += (frameRect.height() * ySpace).toInt()
        }
        var x = 0
        while (x < frameRect.width()) {
            canvas.drawLine(x.toFloat(), 0f, x.toFloat(), frameRect.height().toFloat(), gridPaint)
            x += (frameRect.width() * xSpace).toInt()
        }

        if (mBytes == null) {
            canvas.drawLine(
                0f, waveformRect.height().toFloat(),
                waveformRect.width().toFloat(), waveformRect.height().toFloat(),
                waveformPaint
            )
            return
        }

        if (mPoints == null || mPoints!!.size < mBytes!!.size * 4) {
            mPoints = FloatArray(mBytes!!.size * 4)
        }

        mPoints?.let { mPoints -> mBytes?.let { mBytes ->
                for (i in 0 until mBytes.size - 1) {
                    mPoints[i * 4] =
                        waveformRect.width().toFloat() * i / (mBytes.size - 1).toFloat()
                    mPoints[i * 4 + 1] =
                        waveformRect.height().toFloat() + ((mBytes[i] + sampleRate).toInt()
                            .toByte()) * (waveformRect.height() * peakAmplitude) / sampleRate
                    mPoints[i * 4 + 2] =
                        waveformRect.width().toFloat() * (i + 1) / (mBytes.size.toFloat() - 1)
                    mPoints[i * 4 + 3] =
                        waveformRect.height().toFloat() + ((mBytes[i + 1] + sampleRate).toInt()
                            .toByte()) * (waveformRect.height() * peakAmplitude) / sampleRate
                }
            }
            canvas.drawLines(mPoints, waveformPaint)
        }

    }
}